Support using EOF char (^D) to abort credential prompt

omus opened the pull request:
Part of #20725.

Supports using EOF to abort the credential prompt. This is important as newlines can be used to select the default value.

Note that for password prompts it is currently not possible to tell the difference between an EOF and a newline. The PR behaviour is to assume an EOF was given.

Additional changes include:
- Deprecate `LibGit2.prompt(::AbstractString) -> String` with `Base.prompt(::AbstractString) -> Nullable{String}`
- Lots of tests cases relating to using newline / EOF in the credential prompts (possible now that we have a reliable way of aborting)
- Introduces `LibGit2.user_abort` function which is only meant for internal use.
- An empty SSH username no longer aborts (replaced by EOF abort)
- An empty SSH username_ptr string now triggers a prompt (previously would abort)
- Test function `credential_loop` can now distinguish between "" and C_NULL.

omus committed:
Support using of EOF char (^D) to abort prompt

omus committed:
Prompt when username pointer contains empty string

Without this an empty string in the `username_ptr` will never prompt the
user for a new username. Has not been seen in the wild.

omus committed:
Revise variable name

omus commented:
Note the additional tests produce more of the warnings:
```
WARNING: readuntil(IO,AbstractString) will perform poorly with a long string
```
The warnings will go away once https://github.com/JuliaLang/julia/pull/20621 is merged.

tkelman commented on base/libgit2/utils.jl:
```diff
@@ -37,12 +37,16 @@ function prompt(msg::AbstractString; default::AbstractString="", password::Bool=
+        isempty(uinput) && return Nullable{String}()  # Encountered EOF
+        uinput = chomp(uinput)
+    end
+    Nullable{String}(isempty(uinput) ? default : uinput)
```
is anything externally (PkgDev maybe) using this function that will notice the API change to returning a Nullable?

omus replied:
Good catch. This is used by [PkgDev](https://github.com/JuliaLang/PkgDev.jl/blob/0f8f2b4c3a68572e178c5e479a0d43b826b2d5e6/src/PkgDev.jl#L113). 

What's the recommended procedure for deprecating a function's return type? Should I deprecate `prompt(::AbstractString)` and introduce the following:

```julia
prompt(::Type{Nullable{String}}, msg::AbstractString; default::AbstractString="", password::Bool=false)
```

tkelman replied:
I suppose that'll work, it's not the prettiest or most concise but it would do.

omus replied:
I came up with a better plan. Since this function isn't LibGit2 specific I'll moved this function into `Base`. That way we can have `Base.prompt` and the deprecated `LibGit2.prompt` without having to change the method parameters.

tkelman commented on base/libgit2/callbacks.jl:
```diff
@@ -110,13 +128,16 @@ function authenticate_ssh(creds::SSHCredentials, libgit2credptr::Ptr{Ptr{Void}},
+                    response = prompt("Passphrase for $privatekey", password=true)
+                    isnull(response) && return user_abort()
+                    passdef = unsafe_get(response)
+                    isempty(passdef) && return user_abort()  # Ambiguous if EOF or newline
```
is there still a way to specify / use passwordless keys?

omus replied:
When using `ssh-keygen` a passwordless key wouldn't pass the `passphrase_required` check.

For username/password I don't think LibGit2 will hit this callback unless it requires a password. I'll try to come up with an example for this.

omus replied:
I did some digging into this.

For SSH having a empty passphrase is considered as not having any passphrase at all. Since we can tell the difference between SSH keys that do and do not require passphrases (encrypted private keys have second line as "Proc-Type: 4,ENCRYPTED") and only prompt when a private key is encrypted it should be safe to abort if a empty passphrase is entered at the prompt.

For HTTP I ended up doing some experimentation with [httpbin](https://github.com/kennethreitz/httpbin) and using empty passwords for both HTTP Basic and Digest authentication. Both seem to consider an empty password invalid and will re-prompt for credentials.

tkelman commented on test/libgit2.jl:
```diff
@@ -1549,19 +1578,76 @@ mktempdir() do dir
+                @test auth_attempts == 1
+            end
+
             # TODO: Tests are currently broken. Credential callback prompts for:
```
are you fixing this entire TODO or only part of it?

omus replied:
This TODO is done and should be removed.

tkelman commented on test/libgit2.jl:
```diff
@@ -1549,19 +1578,76 @@ mktempdir() do dir
+            withenv("SSH_KEY_PATH" => nothing,
+                    "SSH_PUB_KEY_PATH" => nothing,
+                    "SSH_KEY_PASS" => nothing,
+                    (Sys.iswindows() ? "USERPROFILE" : "HOME") => tempdir()) do
```
is this relying on tempdir happening to not have any key files in it? probably unlikely, but never know

omus replied:
I specifically need to set the HOME to be a directory that doesn't contain the file ".ssh/id_rsa". I'll add a comment and a test ensuring that this is the case.

omus replied:
I added this check and the comment.

omus committed:
Move LibGit2.prompt to Base.prompt

As part of the supporting EOF update (#23092) we needed to deprecate
the original `LibGit2.prompt` since the return type changed. Since
packages like PkgDev make use of the `prompt` function and the function
itself isn't LibGit2 specific it seemed best to move it into Base. This
allows us to deprecate the original `prompt` without having to change
the function name or modify the parameters.

omus commented:
CI failure looks unrelated. I updated the PR description to mention all changes.

omus commented:
Rebased changes to fix conflicts. Needed to update the LibGit2 test where the user provides an empty private key filename at the prompt (#23108).

```diff
diff --git a/test/libgit2.jl b/test/libgit2.jl
--- a/test/libgit2.jl
+++ b/test/libgit2.jl
@@ -1676,16 +1676,15 @@ mktempdir() do dir
                 @test err == abort_prompt
                 @test auth_attempts == 1

-                # User provides an empty private key
-                # TODO: Should trigger a re-prompt.
+                # User provides an empty private key which triggers a re-prompt
                 challenges = [
                     "Private key location for 'git@github.com':" => "\n",
                     "Public key location for 'git@github.com' [.pub]:" => "\n",
-                    "Passphrase for :" => "\n",
+                    "Private key location for 'git@github.com':" => "\x04",
                 ]
                 err, auth_attempts = challenge_prompt(ssh_cmd, challenges)
                 @test err == abort_prompt
-                @test auth_attempts == 1
+                @test auth_attempts == 2
             end
```

omus commented:
Rebased again to deal with conflicts from #23040. Planning to merge this tomorrow.
